File: doc/test_rooms.cs
#!/usr/bin/env dotnet-script

#r "nuget: WebSocketSharp, 1.0.3-rc11"

using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using WebSocketSharp;

class RoomBroadcastTest
{
    static void Main(string[] args)
    {
        Console.WriteLine("=== Room Broadcast Test ===\n");
        
        var client1 = new WebSocketClient("Client1");
        var client2 = new WebSocketClient("Client2");
        
        Task.Run(() => client1.Connect());
        Task.Run(() => client2.Connect());
        
        Thread.Sleep(1000);
        
        Console.WriteLine("\n[Test] Client1 creating room 'Test Room'...");
        client1.SendMessage(new { type = "CreateRoom", roomName = "Test Room", maxPlayers = 2 });
        
        Thread.Sleep(2000);
        
        Console.WriteLine("\n[Test] Checking if Client2 received the room...");
        Console.WriteLine($"[Test] Client2 rooms count: {client2.RoomsCount}");
        
        if (client2.RoomsCount > 0)
        {
            Console.WriteLine("✅ SUCCESS: Room appeared in Client2!");
        }
        else
        {
            Console.WriteLine("❌ FAIL: Room did NOT appear in Client2!");
        }
        
        Thread.Sleep(100);
    }
}

class WebSocketClient
{
    private WebSocket _ws;
    private string _clientName;
    private List<string> _receivedRooms = new();
    
    public int RoomsCount => _receivedRooms.Count;
    
    public WebSocketClient(string name)
    {
        _clientName = name;
    }
    
    public void Connect()
    {
        _ws = new WebSocket("ws://localhost:5555");
        
        _ws.OnOpen += () =>
        {
            Console.WriteLine($"[{_clientName}] ✅ Connected");
            SendMessage(new { type = "Connect", userId = _clientName, displayName = _clientName });
        };
        
        _ws.OnMessage += (msg) =>
        {
            try
            {
                using (var doc = JsonDocument.Parse(msg))
                {
                    var type = doc.RootElement.GetProperty("Type").GetString();
                    Console.WriteLine($"[{_clientName}] Received: {type}");
                    
                    if (type == "RoomsList")
                    {
                        var rooms = doc.RootElement.GetProperty("Payload").GetProperty("rooms");
                        _receivedRooms.Clear();
                        foreach (var room in rooms.EnumerateArray())
                        {
                            var roomName = room.GetProperty("Name").GetString();
                            _receivedRooms.Add(roomName);
                            Console.WriteLine($"[{_clientName}]   - Room: {roomName}");
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"[{_clientName}] Error parsing message: {ex.Message}");
            }
        };
        
        _ws.OnError += (err) => Console.WriteLine($"[{_clientName}] ❌ Error: {err}");
        _ws.OnClose += () => Console.WriteLine($"[{_clientName}] Disconnected");
        
        _ws.Connect();
    }
    
    public void SendMessage(object message)
    {
        _ws.Send(JsonSerializer.Serialize(message));
    }
}

File: doc/test_rooms.py
#!/usr/bin/env python3
import asyncio
import json
import websockets
import sys

class TestClient:
    def __init__(self, name, url="ws://localhost:5555"):
        self.name = name
        self.url = url
        self.ws = None
        self.rooms = []
        
    async def connect(self):
        try:
            self.ws = await websockets.connect(self.url)
            print(f"[{self.name}] ✅ Connected to server")
            
            # Send Connect message
            connect_msg = {
                "type": "Connect",
                "userId": self.name,
                "displayName": self.name
            }
            await self.ws.send(json.dumps(connect_msg))
            print(f"[{self.name}] Sent Connect message")
            
            # Listen for messages
            asyncio.create_task(self.listen())
            
        except Exception as e:
            print(f"[{self.name}] ❌ Connection error: {e}")
            
    async def listen(self):
        try:
            async for message in self.ws:
                await self.handle_message(message)
        except websockets.exceptions.ConnectionClosed:
            print(f"[{self.name}] Connection closed")
            
    async def handle_message(self, json_str):
        try:
            data = json.loads(json_str)
            msg_type = data.get("Type", "Unknown")
            
            if msg_type == "RoomsList":
                payload = data.get("Payload", {})
                rooms = payload.get("rooms", [])
                self.rooms = rooms
                print(f"[{self.name}] 📍 Received RoomsList with {len(rooms)} room(s)")
                for room in rooms:
                    print(f"[{self.name}]   - {room.get('Name')} ({room.get('Players')}/{room.get('MaxPlayers')})")
            elif msg_type == "RoomCreated":
                print(f"[{self.name}] ✅ Room created")
            elif msg_type == "connected":
                print(f"[{self.name}] ✅ Connection confirmed")
            else:
                print(f"[{self.name}] Received {msg_type}")
        except Exception as e:
            print(f"[{self.name}] Error handling message: {e}")
    
    async def create_room(self, room_name):
        msg = {
            "type": "CreateRoom",
            "roomName": room_name,
            "maxPlayers": 2
        }
        await self.ws.send(json.dumps(msg))
        print(f"[{self.name}] Sent CreateRoom: {room_name}")
    
    async def get_rooms(self):
        msg = {"type": "GetRooms"}
        await self.ws.send(json.dumps(msg))
        print(f"[{self.name}] Requested rooms list")
    
    async def join_room(self, room_id):
        msg = {
            "type": "JoinRoom",
            "roomId": room_id,
            "userId": self.name
        }
        await self.ws.send(json.dumps(msg))
        print(f"[{self.name}] Sent JoinRoom request")

async def main():
    print("=== Room Broadcast & Join Test ===\n")
    
    # Create two clients
    client1 = TestClient("Client1")
    client2 = TestClient("Client2")
    
    # Connect both
    await client1.connect()
    await client2.connect()
    
    # Wait for initial connection and room list
    await asyncio.sleep(1)
    
    # TEST 1: Create room
    print("\n[Test 1] Client1 creating room 'Test Room'...")
    await client1.create_room("Test Room")
    
    # Wait for broadcast
    await asyncio.sleep(2)
    
    print(f"[Test 1] Client2 received {len(client2.rooms)} room(s)")
    
    if len(client2.rooms) > 0:
        print("✅ TEST 1 PASSED: Room appeared in Client2!")
        room = client2.rooms[0]
        print(f"   Room: {room.get('Name')} ({room.get('Players')}/{room.get('MaxPlayers')})")
        
        # TEST 2: Join room
        print("\n[Test 2] Client2 joining room...")
        room_id = room.get('Id')
        await client2.join_room(room_id)
        
        # Wait for join broadcast
        await asyncio.sleep(2)
        
        print(f"[Test 2] Client1 room count: {len(client1.rooms)}")
        if len(client1.rooms) > 0:
            updated_room = client1.rooms[0]
            players_count = updated_room.get('Players')
            print(f"   Updated room: {updated_room.get('Name')} ({players_count}/{updated_room.get('MaxPlayers')})")
            if players_count == 2:
                print("✅ TEST 2 PASSED: Room now shows 2 players on Client1!")
            else:
                print(f"❌ TEST 2 FAILED: Expected 2 players, got {players_count}")
        else:
            print("❌ TEST 2 FAILED: No rooms on Client1")
    else:
        print("❌ TEST 1 FAILED: Room did NOT appear in Client2!")

if __name__ == "__main__":
    asyncio.run(main())
